library/nmstate_linkagg.py

```diff
@@ -0,0 +1,239 @@
+  net_linkagg:
+    aggregate:
+      - name: bond0
+      - name: bond1
+    state: absent
+'''
+
+RETURN = '''
+state:
+    description: Network state after running the module
+    type: dict
+'''
+
+
+def run_module():
+    element_spec = dict(
+        members=dict(type='list'),
+        min_links=dict(type='int'),
+        # net_linkagg only knows on, active and passive
+        # on and active is 802.3ad on Linux, passive is not supported
+        mode=dict(choices=['on', 'active', 'passive', 'balance-rr',
+                           'active-backup', 'balance-xor', 'broadcast',
+                           '802.3ad', 'balance-tlb', 'balance-alb'],
+                  default='on'),
+        name=dict(),
+        state=dict(default='present',
+                   choices=['present', 'absent', 'up', 'down'])
+    )
+
+    aggregate_spec = deepcopy(element_spec)
+    aggregate_spec['name'] = dict(required=True)
+
+    # remove default in aggregate spec, to handle common arguments
+    remove_default_spec(aggregate_spec)
+
+    argument_spec = dict(
+        aggregate=dict(type='list', elements='dict', options=aggregate_spec),
+        purge=dict(default=False, type='bool'),
+        # not in net_* specification
+        debug=dict(default=False, type='bool'),
+    )
+
+    argument_spec.update(element_spec)
+
+    required_one_of = [['name', 'aggregate']]
+    mutually_exclusive = [['name', 'aggregate']]
+
+    result = dict(
+        changed=False,
+    )
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        required_one_of=required_one_of,
+        mutually_exclusive=mutually_exclusive,
+        supports_check_mode=True
+    )
+
+    previous_state = netinfo.show()
+
+    if module.params['aggregate']:
+        module.fail_json(msg='Aggregate not yet supported', **result)
+
+    # leader interface
+    members = module.params['members']
+    if not isinstance(members, list):
+        members = [members]
+    mode = module.params['mode']
+    if mode in ['on', 'active']:
+        mode = '802.3ad'
+    elif mode in ['passive']:
+        # passive mode is not supported on Linux:
+        # noqa:
+        # https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/networking_guide/sec-comparison_of_network_teaming_to_bonding
+        module.fail_json(msg='passive mode is not supported on Linux',
+                         **result)
+
+    link_aggregation = {'mode': mode,
+                        'options': {},  # FIXME: add support for options?
+                        'slaves': members,
+                        }
+    interface_state = {'name': module.params['name'],
+                       'state': module.params['state'],
+                       'type': 'bond',
+                       'link-aggregation': link_aggregation,
+                       }
+    interfaces = []
+
+    # only add info about members when they are mentioned
+    if module.params['state'] in ['up', 'present'] and False:
```

Comment:
Isn't "and False" at the end makes this expression always resolve to False?
Looks like dead code.

Reply:
I removed this and added the right code to fail when members are missing in a later commit.